importing pyrms and diffeqpy during a debug session

Laxzal opened the pull request:
<!--
Thanks for contributing a pull request! Please try to provide as much detail as possible to help the reviewer understand your work.
You can also add the appropriate labels to describe the topic of the pull request and the type of changes you're making.
-->

### Motivation or Problem
Updated the import of Julia so that now it will first check if it is in debug mode. If so, will import Julia() first prior to import pyrms/diffeqpy. Furthermore, during debug mode, will look for system image of RMS is exists. Refer to #2363 

### Description of Changes
I have edited `reactors.py` to now determine if the user is debugging. If the user is not debugging, it will follow the original code, trying importing pyrms/diffeqpy. If exception is met, it will pass (this is obviously important for testing in ARC and T3).

However, if the user is debugging, it will attempt to get the top path of the `reactors.py` file and then search for a system image of `rms.so`. If not found, it will load Julia without precompiled modules. This allows for pyrms and diffeqpy to be imported.

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with main?
-->

alongd reviewed (commented):
Overall looks good to me

I could suggest to begin with `if gettrace()`, and then regardless of the value of `gettrace` run the lines that were deleted here. Just a minor suggestion to avoid repetition

mjohnson541 commented:
Looks good to me as well!

Laxzal commented:
It seems the CI/RMG-Tests is failing during the compiling of RMS

```console
 ERROR: Unsatisfiable requirements detected for package StochasticDiffEq [789caeaf]:
 StochasticDiffEq [789caeaf] log:
 ├─possible versions are: 5.0.0-6.57.4 or uninstalled
 ├─restricted to versions 6.36.0 by an explicit requirement, leaving only versions 6.36.0
 └─restricted by compatibility requirements with OrdinaryDiffEq [1dea7af3] to versions: [5.0.0-6.20.0, 6.43.0-6.57.4] or uninstalled — no versions left
   └─OrdinaryDiffEq [1dea7af3] log:
     ├─possible versions are: 4.0.0-6.40.1 or uninstalled
     └─restricted to versions 6 by ReactionMechanismSimulator [c2d78dd2], leaving only versions 6.0.0-6.40.1
       └─ReactionMechanismSimulator [c2d78dd2] log:
         ├─possible versions are: 0.4.0 or uninstalled
         └─ReactionMechanismSimulator [c2d78dd2] is fixed to version 0.4.0
```

Laxzal commented:
I am getting an fail in the ci/rmg_tests. During a test, it says cannot find 'rms'
And looking at the install of RMS, I see this

```console
ERROR: Unsatisfiable requirements detected for package StochasticDiffEq [789caeaf]:
 StochasticDiffEq [789caeaf] log:
 ├─possible versions are: 5.0.0-6.57.4 or uninstalled
 ├─restricted to versions 6.36.0 by an explicit requirement, leaving only versions 6.36.0
 └─restricted by compatibility requirements with OrdinaryDiffEq [1dea7af3] to versions: [5.0.0-6.20.0, 6.43.0-6.57.4] or uninstalled — no versions left
   └─OrdinaryDiffEq [1dea7af3] log:
     ├─possible versions are: 4.0.0-6.41.0 or uninstalled
     └─restricted to versions 6 by ReactionMechanismSimulator [c2d78dd2], leaving only versions 6.0.0-6.41.0
       └─ReactionMechanismSimulator [c2d78dd2] log:
         ├─possible versions are: 0.4.0 or uninstalled
         └─ReactionMechanismSimulator [c2d78dd2] is fixed to version 0.4.0
```

mjohnson541 commented:
No idea why this popped up now instead of earlier, but this seems to be the result of us pinning the StochasticDiffEq.jl version in the RMG-tests install script at some point. This isn't done in the latest CI build and documentation build scripts, which run successfully. Removing the part of the lines here: https://github.com/ReactionMechanismGenerator/RMG-tests/blob/10da9f5656baeb31091cb287c232b4f9a0aca7e4/install.sh#L114 that install a specific version of StochasticDiffEq.jl should fix the build.

Laxzal commented:
@mjohnson541 Okay, so I do you reckon I should make the changes in this branch or make another branch for that change regarding the unpinning of StoachasticDiffEq

mjohnson541 commented:
It's interesting that all RMG-tests runs aren't failing: https://github.com/ReactionMechanismGenerator/RMG-tests/actions/runs/4060446913/jobs/6989541587

Laxzal commented:
Appears I don't have permissions to push a branch to RMG-tests.

But yeah, it is concerning to me that my change is having a failure during the installation of RMS. I can't imagine how my code change affects in particular

Laxzal commented:
@mjohnson541 I checked the link and it also has the same issue
```julia
ERROR: Unsatisfiable requirements detected for package StochasticDiffEq [789caeaf]:
 StochasticDiffEq [789caeaf] log:
 ├─possible versions are: 5.0.0-6.57.4 or uninstalled
 ├─restricted to versions 6.36.0 by an explicit requirement, leaving only versions 6.36.0
 └─restricted by compatibility requirements with OrdinaryDiffEq [1dea7af3] to versions: [5.0.0-6.20.0, 6.43.0-6.57.4] or uninstalled — no versions left
   └─OrdinaryDiffEq [1dea7af3] log:
     ├─possible versions are: 4.0.0-6.41.0 or uninstalled
     └─restricted to versions 6 by ReactionMechanismSimulator [c2d78dd2], leaving only versions 6.0.0-6.41.0
       └─ReactionMechanismSimulator [c2d78dd2] log:
         ├─possible versions are: 0.4.0 or uninstalled
         └─ReactionMechanismSimulator [c2d78dd2] is fixed to version 0.4.0
```
So I reckon that my change is not letting the pass occur anymore? It is trying to import rms from pyrms and then failing there...

mjohnson541 commented:
You should have permissions now.

Laxzal commented:
Yeah so I am a bit confused why my changes suddenly throws the rms error.

Is RMG-tests based upon the tests for RMG-Py? I am just wondering how I could debug it to see where the issue is occuring

mjohnson541 commented:
Hopefully that will resolve the tests issue. RMG-tests runs a set of RMG jobs and looks at how estimation, species selection and some predictions have changed.

alongd commented:
@Laxzal , can you try rebasing to fix the branch's history? I can help offline if needed

Laxzal commented:
@alongd It should be fixed now

@mjohnson541 
I am unsure if this is because of the removal of the version pin but RMG-tests now gets this error

```console
ERROR: InitError: PyError (PyImport_ImportModule
The Python package rmgpy.molecule could not be imported by pyimport. Usually this means
that you did not install rmgpy.molecule in the Python version being used by PyCall.
PyCall is currently configured to use the Python version at:
/usr/share/miniconda/envs/testing/bin/python
and you should use whatever mechanism you usually use (apt-get, pip, conda,
etcetera) to install the Python package containing the rmgpy.molecule module.
One alternative is to re-configure PyCall to use a different Python
version on your system: set ENV["PYTHON"] to the path/name of the python
executable you want to use, run Pkg.build("PyCall"), and re-launch Julia.
Another alternative is to configure PyCall to use a Julia-specific Python
distribution via the Conda.jl package (which installs a private Anaconda
Python distribution), which has the advantage that packages can be installed
and kept up-to-date via Julia.  As explained in the PyCall documentation,
set ENV["PYTHON"]="", run Pkg.build("PyCall"), and re-launch Julia. Then,
To install the rmgpy.molecule module, you can use `pyimport_conda("rmgpy.molecule", PKG)`,
where PKG is the Anaconda package that contains the module rmgpy.molecule,
or alternatively you can use the Conda package directly (via
`using Conda` followed by `Conda.add` etcetera).
) <class 'ModuleNotFoundError'>
ModuleNotFoundError("No module named 'rmgpy.rmgobject'")
  File "/home/runner/work/RMG-tests/RMG-tests/code/testing/RMG-Py/rmgpy/molecule/__init__.py", line 31, in <module>
    from rmgpy.molecule.element import Element, PeriodicSystem, get_element
  File "/home/runner/work/RMG-tests/RMG-tests/code/testing/RMG-Py/rmgpy/molecule/element.py", line 47, in <module>
    from rmgpy.quantity import Quantity
  File "/home/runner/work/RMG-tests/RMG-tests/code/testing/RMG-Py/rmgpy/quantity.py", line 43, in <module>
    from rmgpy.rmgobject import RMGObject, expand_to_dict
Stacktrace:
 [1] pyimport(name::String)
   @ PyCall /usr/share/miniconda/envs/testing/share/julia/site/packages/PyCall/twYvK/src/PyCall.jl:558
 [2] pyimport_conda(modulename::String, condapkg::String, channel::String)
   @ PyCall /usr/share/miniconda/envs/testing/share/julia/site/packages/PyCall/twYvK/src/PyCall.jl:716
 [3] __init__()
   @ ReactionMechanismSimulator /usr/share/miniconda/envs/testing/share/julia/site/packages/ReactionMechanismSimulator/eFaj4/src/ReactionMechanismSimulator.jl:22
 [4] _include_from_serialized(path::String, depmods::Vector{Any})
   @ Base ./loading.jl:696
 [5] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String)
   @ Base ./loading.jl:782
 [6] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:1020
 [7] require(uuidkey::Base.PkgId)
   @ Base ./loading.jl:936
 [8] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:923
during initialization of module ReactionMechanismSimulator
```
Could be the LD_LIBRARY_PATH issue?

Laxzal commented:
Looks like the PR passed. I think the changes to the debugging checked fixed it? I am not certain though

mjohnson541 commented:
Yeah...I think there's still an issue on RMG-tests, it can probably be fixed just by making the install more and more like the CI builds until it's fixed, but since we aren't using RMS reactors RMG-tests seems to be able to avoid breaking, which is enough for this PR. 

Can you squash and write an informative commit message?

Laxzal commented:
> Yeah...I think there's still an issue on RMG-tests, it can probably be fixed just by making the install more and more like the CI builds until it's fixed, but since we aren't using RMS reactors RMG-tests seems to be able to avoid breaking, which is enough for this PR.
> 
> Can you squash and write an informative commit message?

No worries. I have squashed and updated the commit message

kfir4444 reviewed (approved):
Looks great @Laxzal
